assert/assertions.go

```diff
@@ -75,46 +75,75 @@ func ObjectsAreEqual(expected, actual interface{}) bool {
 	return bytes.Equal(exp, act)
 }
 
-// ObjectsExportedFieldsAreEqual determines if the exported (public) fields of two structs are considered equal.
-// If the two objects are not of the same type, or if either of them are not a struct, they are not considered equal.
-//
-// This function does no assertion of any kind.
-func ObjectsExportedFieldsAreEqual(expected, actual interface{}) bool {
-	if expected == nil || actual == nil {
-		return expected == actual
+// copyExportedFields iterates downward through nested data structures and creates a copy
+// that only contains the exported struct fields.
+func copyExportedFields(expected interface{}) interface{} {
+	if isNil(expected) {
+		return expected
 	}
 
 	expectedType := reflect.TypeOf(expected)
-	actualType := reflect.TypeOf(actual)
-
-	if expectedType != actualType {
-		return false
-	}
-
-	if expectedType.Kind() != reflect.Struct || actualType.Kind() != reflect.Struct {
-		return false
-	}
-
+	expectedKind := expectedType.Kind()
 	expectedValue := reflect.ValueOf(expected)
-	actualValue := reflect.ValueOf(actual)
-
-	for i := 0; i < expectedType.NumField(); i++ {
-		field := expectedType.Field(i)
-		isExported := field.PkgPath == "" // should use field.IsExported() but it's not available in Go 1.16.5
-		if isExported {
-			var equal bool
-			if field.Type.Kind() == reflect.Struct {
-				equal = ObjectsExportedFieldsAreEqual(expectedValue.Field(i).Interface(), actualValue.Field(i).Interface())
-			} else {
-				equal = ObjectsAreEqualValues(expectedValue.Field(i).Interface(), actualValue.Field(i).Interface())
+
+	switch expectedKind {
+	case reflect.Struct:
+		result := reflect.New(expectedType).Elem()
+		for i := 0; i < expectedType.NumField(); i++ {
+			field := expectedType.Field(i)
+			isExported := field.PkgPath == "" // should use field.IsExported() but it's not available in Go 1.16.5
```

Comment:
We probably shouldn't support Go 1.16 any more since it's not officially supported anyways.

Reply:
Feel free to change the build to only support 1.20, 1.19, and 1.18 -- you can then change this code to not consider Go 1.16

Reply:
@boyan-soubachov Updated!

Be aware the the formatting the auto-formatting rules seem to have changed for versions >= 1.19. So we are now skipping that part of the build job for 1.18.